rust/swig/typedb_client_python.swg

```diff
@@ -0,0 +1,112 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/* Arrays handling */
+%define %array(Type)
+%typemap(in) Type ** (int size) {
+    size = PyList_Size($input);
+    $1 = (Type **) malloc((size + 1) * sizeof(Type *));
+    int i = 0;
+    for (i = 0; i < size; ++i) {
+        void *argp = 0;
+        int res = SWIG_ConvertPtr(PyList_GetItem($input, i), &argp, SWIGTYPE_p_##Type, 0);
+        if (!SWIG_IsOK(res)) {
+            SWIG_exception_fail(SWIG_ArgError(res), "array item is not of type 'Type const *'");
+          }
+        $1[i] = reinterpret_cast<Type *>(argp);
+    }
+    $1[i] = 0;
+}
+
+%typemap(freearg) Type ** { free($1); }
+%enddef
+
+%array(Annotation)
+%array(Concept)
+%array(Thing)
+
+/* char** needs special handling */
+%typemap(in) char ** (int size) {
+    if (PyList_Check($input)) {
+        size = PyList_Size($input);
+        int i = 0;
+        $1 = (char **) malloc((size + 1) * sizeof(char *));
+        for (i = 0; i < size; ++i) {
+            PyObject *obj = PyList_GetItem($input, i);
+            if (PyUnicode_Check(obj)) {
+                const char *str = PyUnicode_AsUTF8(PyList_GetItem($input, i));
+                $1[i] = (char *)malloc((strlen(str) + 1) * sizeof(char));
+                strcpy($1[i], str);
+            }
+            else {
+                PyErr_SetString(PyExc_TypeError, "list must contain strings");
+                for (int j = 0; j < i; ++j)
+                        free($1[j]);
+                free($1);
+                return NULL;
+            }
+        }
+        $1[i] = 0;
+    } else {
+        PyErr_SetString(PyExc_TypeError, "not a list");
+        return NULL;
+    }
+}
+
+%typemap(freearg) char ** {
+    for (int i = 0; i < size$argnum - 1; ++i)
+        free($1[i]);
+    free((char *) $1);
+}
+
+/* Introduce TypeDBClientException and use it to wrap all native exceptions */
+%{
+static PyObject* PyExc_TypeDBClientError;
+%}
+
+%init %{
+    PyExc_TypeDBClientError = PyErr_NewException("native_client_python.TypeDBClientException", NULL, NULL);
+    Py_INCREF(PyExc_TypeDBClientError);
+    PyModule_AddObject(m, "TypeDBClientException", PyExc_TypeDBClientError);
+%}
+
+%exception {
+    $action
+    if (check_error()) {
+        PyErr_SetString(PyExc_TypeDBClientError, error_message(get_last_error()));
+        SWIG_fail;
+    }
+}
+
+%pythoncode %{
+    TypeDBClientException = native_client_python.TypeDBClientException
+%}
+
+/* ValueType* needs special handling */
+%typemap(in) enum ValueType * {
+    if ($input != Py_None) {
```

Comment:
How does swig translate `none`? Is this the protection we built for the issue we had found where we passed a null and it got mapped to `object` implicitly?

Reply:
The main problem was with all values except `None`. Now we allocate memory for all such `ValueType`s (and free it later), and the value remains `null` if it was `None` in python.

Reply:
ok, got it. i hope we don't introduce any memory leaks!!